Moved RepositoryParamProvider to ResourceEntry

ieugen opened the pull request:
* rmeoved from controller
* cleaner API

meshuga commented:
I don't think it's a good change. Take a look at `ServletParametersProvider` in `katharsis-servlet`. This (and every other `ParameterProvider`) has to have a reference to a request.

ieugen commented:
Yes, I made some changes after I integrated my work with katharsis-vertx. I updated the code and I added it to the Request object we pass to the framework. 
I'm refactoring tests and so far there are 13 failing tests. 

After I finish I'll make a PR with all of the commits and an internal release of both core and vertx. We will deploy the new version it to staging tomorrow and come back with fixes.

ieugen commented:
We have working integration testing via kotlin itests :).

chb0github commented on pom.xml:
```diff
@@ -54,20 +58,66 @@
                 <configuration>
                     <instructions>
-                        <Export-Package>io.katharsis.dispatcher.*,io.katharsis.errorhandling.*,io.katharsis.jackson.*,io.katharsis.locator.*,io.katharsis.queryParams.*,io.katharsis.repository.*,io.katharsis.request.*,io.katharsis.resource.*,io.katharsis.response.*,io.katharsis.utils.*</Export-Package>
+                        <Export-Package>
```
Why? What does this buy you?

ieugen replied:
Readability.

chb0github commented on pom.xml:
```diff
@@ -54,20 +58,66 @@
+            <artifactId>commons-vfs2</artifactId>
+            <version>2.0</version>
+            <scope>provided</scope>
+            <optional>true</optional> <!-- case: -->
```
Can you explain your thinking on this dependency? And, exactly why it's may/maynot be there and "that's ok".

ieugen replied:
It is transitive dependency that did not show down the road (in aps using katharsis). I added it to help maven figure things out.

chb0github commented on pom.xml:
```diff
@@ -93,6 +143,45 @@
             <scope>test</scope>
         </dependency>
         <dependency>
+            <groupId>org.jetbrains.kotlin</groupId>
```
You know, I am not opposed to using a vendor-specific language (even if open sourced) but you should get group input on this before injecting a bunch of code. Mixing up JVM languages in the same workspace is rare. @meshuga @masterspambot  thoughts?

ieugen replied:
I've limited the code only in tests. Maven does not help in providing a good separation. Gradle is better. Migrating to gradle will help us separate source trees better. Kotlin classes are kept as tests only and are for the purpose of evaluating Kotlin. So far, it proved good from my point of view. A lot less code to write when doing tests.

chb0github commented on pom.xml:
```diff
@@ -118,6 +207,10 @@
         </repository>
+        <repository>
+            <id>jebrains-all</id>
+            <url>http://repository.jetbrains.com/all</url>
```
The kotlin libs aren't published to Maven central?

ieugen replied:
I think I was experimenting with some libraries and fogot that repo there.

chb0github commented on src/main/java/io/katharsis/dispatcher/JsonApiDispatcherImpl.java:
```diff
@@ -0,0 +1,97 @@
+ * and katharsis-servlet for usage.
+ */
+@Data
+public class JsonApiDispatcherImpl implements JsonApiDispatcher {
```
This is an antipattern.

Use of `Impl` in a class name doesn't serve any purpose. Something more like `DefaultJsonApiDispatcher` is more appropriate.

chb0github commented on src/main/java/io/katharsis/dispatcher/JsonApiDispatcher.java:
```diff
@@ -0,0 +1,10 @@
+ * Dispatches JSON API requests to proper handlers.
+ * (ieugen: seems to be identical to a handler so far)
+ */
+public interface JsonApiDispatcher extends JsonApiHandler {
```
And interface with no methods?

ieugen replied:
I was experimenting with ways of building the API. dispatcher and handler seem to have the same interface. Until things cleared up I decided to inherit from one another.

chb0github commented on src/main/java/io/katharsis/dispatcher/JsonApiDispatcherImpl.java:
```diff
@@ -0,0 +1,97 @@
+        } catch (KatharsisMatchingException ke) {
+            response.setHttpStatus(406);
+        } catch (Exception e) {
+            response = toErrorResponse(e, 422);
```
This is is a bug. You're going to catch your own `MethodNotFoundException` and return a 422? 422 is a WebDav error and has no application at all. This should throw an `UnsupportedOperationException` and result in a 405: `METHOD_NOT_ALLOWED`

Also, what about `OPTIONS` and `HEAD` ?
If you look at the spec for for 405 `HEAD` should never report an error.

ieugen replied:
Thanks for the info. 

Please keep in mind that the whole re-write is NOT FINISHED. There are howver a lot of architectural changes that clear things up. I will take some time to document them so it is easy to follow. 

Regarding 422, github seems to use it for this purpose and it is there for that meaning so why not use it? It's not forbidden and it is more clear in explainig what happened.

OPTIONS and HEAD are not supported right now. 

[1] https://developer.github.com/v3/
[2] http://stackoverflow.com/questions/16133923/400-vs-422-response-to-post-of-data

chb0github commented on src/main/java/io/katharsis/dispatcher/JsonApiDispatcherImpl.java:
```diff
@@ -0,0 +1,97 @@
+        ResponseContext errorResponse = null;
+
+        if (exceptionMapper.isPresent()) {
+//            errorResponse = exceptionMapper.get().toErrorResponse(e);
```
Why is this code even here?

As an asside: An ExceptionMapper should always exist, even for `Exception` -- is present should be replaced with `true` -- if that condition doesn't hold it's an error

ieugen replied:
Please keep in mind that the whole re-write is NOT FINISHED.

chb0github commented on src/main/java/io/katharsis/dispatcher/JsonApiDispatcherImpl.java:
```diff
@@ -0,0 +1,97 @@
+             * No extra processing needs to be done - body parsing, etc.
+             */
+
+            switch (request.getMethod()) {
```
This logic should be switched  and turned into a static listing of supported values:

```

private static Set<Method> SUPPORTED_METHODS = newHashSet<>(Arrays.asList(GET,POST,PATCH,DELETE));

if(!SUPPORTED_METHODS.contains(request.getMethod())
   throw new UnsupportedOperationException("Currently not supported");
```
Less code, less Cyclomatic complexity, less indenting. 15 lines of code down to 3

ieugen replied:
Nice one. Thanks.

chb0github commented:
A complete rewrite of the internals of the system needs a commit-by-commit review. There is way too much to review here and what I have reviewed I have supplied feedback on.

I suggest opening an issue to discuss the refactor and then doing many pull requests.

ieugen commented:
I agree that things are not nice. I got here by starting to do a small refactor that turned into a very big one. 

My approach will be to finish the rewrite until we have an enough stable API and we can replace the old version with the new. After that we can create a2.x maintanence branch and merge this PR into develop. That way we can maintaing 2.4.x for some time and still be able to work on the new version. 

I will document the design decisions on Read the docs site [1] and keep you posted. 

Again, the PR is not ready but getting closer. 

[1] http://katharsis-jsonapi.readthedocs.io/en/latest/

Ramblurr commented:
@ieugen Just let me and @weaselmetal know when it is ready for testing.
